### DISCUSSION TITLE
How to implement distributed systems

### DISCUSSION THREAD

**Discussion by liusuyi2021** [bluenviron/mediamtx#2724] [General]
How can I achieve distributed deployment, and when a single mediaMTX reaches its limit, I want to scale the service?

**Comment by aler9** [comment]
Hello, while native scalability is still under development (see #998), you can already instantiate as many instances as you want and distribute the load by using a TCP load balancer in front of them. This works perfectly until all instances have the same configuration and all the sources are fixed (i.e. they point to fixed endpoints):

```
user --> load balancer --> instance 1
                       --> instance 2
```

On the other hand, if sources are dynamic (i.e. they are published by a client), you have to use a single instance and distribute the load by using a swarm of other secondary instances in front of it, each configured to proxy requests from the main instance:

```
user --> load balancer --> instance 1 --> main instance
                       --> instance 2
```

**Reply by joshefin** [reply]
@aler9 I would like to use MediaMTX to ingest multiple dynamic streams over RTMP and deliver them to many users over LL-HLS. A single instance won't be enough. Looking at the documentation and discussions, even though scalability is not a supported feature, could this architecture be a good way to go?

```
                                                         ----------> 
                            --> [mediamtx ingest node 1] ----------> 
users --(rtmp)--> [haproxy] --> [mediamtx ingest node 2] --(rtsp)--> ...
                            --> [mediamtx ingest node 3] ----------> 
                                                         ----------> 

    [mediamtx delivery node 1] ----------->
    [mediamtx delivery node 2] ----------->
... [mediamtx delivery node 3] --(llhls)--> [nginx] ---> users
    [mediamtx delivery node 4] ----------->
    [mediamtx delivery node 5] ----------->

```
- HAProxy for load balancing incoming RTMP
- Ingest nodes to handle the incoming RTMP streams
- Delivery nodes that pull streams from ingest nodes over RTSP, then generate and serve LL-HLS files
- nginx to load balance HTTP to delivery nodes and optionally cache files

This requires a mechanism so each delivery node knows which ingest node holds a given stream. One idea is to use runOn* hooks in ingest-node onfiguration to invoke a script that stores stream and node names in Redis. Delivery nodes could then query Redis to determine the correct RTSP URL for each stream - I'm just not sure how to implement this part.

Is this feasible in the current state of MediaMTX?

**Reply by aler9** [reply]
@joshefin the architecture you described is what i would call "stateful balancing", because there's a Redis server that keeps tracks of which delivery node is in charge of handling a certain stream, but there's also a simpler alternative, which is "stateless balancing", in which you configure ingest nodes in which way:

```yml
paths:
  '~^(a)':
    runOnReady: forward or redirect stream to delivery node 1
  '~^(b)':
    runOnReady: forward or redirect stream to delivery node 2
...
```

Basically you are redirecting streams that starts with the letter "a" to node 1, streams that starts with letter "b" to node 2, and so on. If you think about it, it's what most HTTP load balancers do (round robin), with the difference that they use a random seed to identify the delivery node, while here i'm using stream names.
